### ISSUE TITLE
Support infinite streams

### ISSUE THREAD

**Issue by udoprog** [rust-lang-nursery/futures-rs#1857]
This is an issue opened to consider adding support for infinite streams.

Infinite streams are streams which do not terminate, so they yield values of `T` directly instead of `Option<T>`. This is useful for things which conceptually never ends, like `Interval`s, channels, or lazy settings updates. The fact that we currently have a panicking `select_next_some` speaks to me that there might be space for this.

Infinite streams would look something like this:

```rust
trait InfiniteStream {
    type Item;

    fn poll_next(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Self::Item>;
}
```

Or in a perfect world, consider making the existing `Stream` not resolve with an `Option<T>` and make terminating streams a specialization on this. Similarly to how `Result<T, E>` was removed from `Future`.

The one sticking point I'm not sure about is if combinators could be implemented once instead of once per stream variant (terminating vs infinite).

CC: #1765

**Comment by canndrew** [comment]
`Stream` should have two associated items: `Yield` and `Return`. That way you can have:

* The current stream behaviour

        Stream<Yield = T, Return = ()>

* Infinite streams

        Stream<Yield = T, Return = !>

* infinite streams that have fatal errors

        Stream<Yield = T, Return = E>

* finite streams that have both fatal and non-fatal errors and allow you to distinguish them

        Stream<Yield = Result<T, NonFatalE>, Return = Result<(), FatalE>>

* etc.

**Comment by udoprog** [comment]
> `Stream` should have two associated items: `Yield` and `Return`. That way you can have:

I can't help but feel this reasoning is similar to what lead to `Future` polling to `Poll<Self::Item, Self::Error>`, which is an alias for `Result<Async<Self::Item>, Self::Error>`. I fear that kind of specialization mainly hurts us in the long term since it tends to color the entire ecosystem. It's also worth considering that `never_type` is not yet stable.

Is there a concrete benefit with this, over specializing over an enum like `YieldOrReturn` (which also happens to be what [`GeneratorState`](https://doc.rust-lang.org/std/ops/enum.GeneratorState.html) is) similarly to what `TryFuture` and `TryStream` does now?

**Comment by Ralith** [comment]
I don't think this is quite the same situation as `Future`'s historical `Error` type, because whether a `Future` returned `Ok(Ready(_))` or `Err(_)` did not confer any information about the internal state of the future--it's completed regardless--whereas returning `Yield` or `Return` distinguishes whether the `Stream` may still be polled. That said, I think the differences between these two approaches are mostly cosmetic, so maybe it would be instructive to look at a sampling of combinators defined using each, and some uses of them, and see how they compare?

**Comment by nbdd0121** [comment]
I am curious about the impact of the `Stream` design on potential future `for await` syntax, since `Stream` is supposed to be an async version of `Iterator`, isn't it?

**Comment by seanmonstar** [comment]
Relevant previous discussion: https://github.com/rust-lang-nursery/futures-rs/issues/1059

**Comment by Ralith** [comment]
Yeah, that use-case would still work just fine if it was an infinite stream that had a sentinel value like `Err(Terminated)`. In Quinn I currently have some streams that yield `Some(Err(_))` forever rather than returning `None` because I assume readers will bail out on encountering an error and going to extra effort to do something else after returning a terminal value feels strange, but this has already bitten me in test code that relied abstractly on a `None` eventually arriving. Having *any* sort of more flexible pattern for a termination value would have avoided this confusion.

The dwindling importance of combinators is a valid point in some respects, but if you specifically need to transform a stream to pass it on to another API, combinators remain the only option at least until we get something like async generators. There's also no easy way to supplant e.g. `for_each_concurrent`.

**Comment by Nemo157** [comment]
> But mimicking what `std::sync::channel` does with [`RecvError`](https://doc.rust-lang.org/std/sync/mpsc/struct.RecvError.html) might be clearer from a documentation perspective and having a stronger correlation between the APIs.

[`std::sync::mpsc::IntoIter`](https://doc.rust-lang.org/std/sync/mpsc/struct.IntoIter.html) has the same behaviour of turning a disconnection into end-of-iterator. The main difference here seems to just be that `futures::channel::mpsc::Receiver` doesn't have a standalone `async fn recv(&self) -> ?` and requires you to go via the `Stream` impl.

**Comment by Nemo157** [comment]
>  I can't actually recall from the top of my head a stream implementation which is finite that I've used during the last couple of months. Note that there's plenty that wraps a `Result` for _fallible_ streams though which terminate for other reasons.

I've had the opposite experience, the majority of my `Stream` usage has been wrappers around multi-message IO channels, which all gracefully terminate when disconnected. The next major part has been `mpsc` channels which are in some way connected to an IO channel, so will terminate when the IO terminates. The only infinite streams I can think of using are simple timer ticks, and the `Incoming` stream from a `TcpListener`.

**Comment by udoprog** [comment]
> I've had the opposite experience, the majority of my `Stream` usage has been wrappers around multi-message IO channels, which all gracefully terminate when disconnected.

Hm. Do these tend to resolve to `Option<Result<T, E>>`, or just `Option<T>`? I mostly spend my time writing high-level async code. Things that should run until the application stops (or errors) and the futures are dropped.

**Comment by Nemo157** [comment]
They're mostly per-client `Stream<Item = Result<T, E>>`, they could have an IO error (e.g. something in the network connection fails) or a parse error (e.g. the other side sends an invalid message), which would normally result in me just abandoning that task and dropping all other connections it's dealing with (all associated to that one client). If the other side closes the connection the the stream will cleanly finish and I instead close all other connections and wait for them to complete before the task finishes.
